src/main/java/org/openstreetmap/atlas/checks/validation/linear/lines/WaterWayCheck.java

```diff
@@ -0,0 +1,526 @@
+    {
+        return FALLBACK_INSTRUCTIONS;
+    }
+
+    /**
+     * Create a flag for an object that goes uphill
+     *
+     * @param flag
+     *            The flag to create/modify. May be {@code null}.
+     * @param object
+     *            The object to flag
+     * @param first
+     *            The first node of the way
+     * @return The new CheckFlag (if the passed flag was {@code null}) or the modified CheckFlag.
+     */
+    private CheckFlag createUphillFlag(final CheckFlag flag, final AtlasObject object,
+            final Location first)
+    {
+        final CheckFlag returnFlag = flag;
+        final String instruction = this.getLocalizedInstruction(
+                FALLBACK_INSTRUCTIONS.indexOf(GOES_UPHILL), object.getOsmIdentifier(),
+                this.elevationUtils.getResolution(first).asMeters());
+        if (returnFlag == null)
+        {
+            return createFlag(object, instruction);
+        }
+        returnFlag.addInstruction(instruction);
+        return returnFlag;
+    }
+
+    /**
+     * Check if a line crosses an coastline line
+     *
+     * @param line
+     *            The line to check
+     * @return {@code true} if the line crosses a coastline
+     */
+    private boolean doesLineCrossCoast(final LineItem line)
+    {
+        final List<LineItem> lines = new ArrayList<>();
+        line.getAtlas()
+                .lineItemsIntersecting(line.asPolyLine().bounds(), this.oceanBoundaryTags::test)
+                .forEach(lines::add);
+        return !lines.isEmpty();
+    }
+
+    /**
+     * Check if the waterway ends in a sink (i.e., a location that can reasonably expected to have
+     * no outflow).
+     *
+     * @param line
+     *            The waterway item to check
+     * @return {@code true} if the waterway ends in a sink.
+     */
+    private boolean doesWaterwayEndInSink(final LineItem line)
+    {
+        return doesLineEndOnWaterway(line) || doesLineEndInSink(line) || doesLineEndInOcean(line);
+    }
+
+    /**
+     * Get an intersecting waterway, if one exists.
+     *
+     * @param line
+     *            The waterway to look for intersections for.
+     * @return A crossing waterway, or {@code null} if no crossing waterway exists.
+     */
+    private LineItem intersectsAnotherWaterWay(final Atlas atlas, final LineItem line)
+    {
+        final Iterable<LineItem> intersectingWaterways = atlas.lineItemsIntersecting(line.bounds(),
+                this.waterwayTagFilter::test);
+        final Set<LineItem> sameLayerWays = Iterables.stream(intersectingWaterways)
+                .filter(potential -> LayerTag.areOnSameLayer(line, potential)
+                        && !waterwayConnects(line, potential))
+                .collectToSet();
+        sameLayerWays.removeIf(line::equals);
+        if (sameLayerWays.isEmpty())
+        {
+            return null;
+        }
+        return sameLayerWays.iterator().next();
```

Comment:
I was trying to "flag" each object as quickly as possible, to avoid significantly increasing processing times. Since then, I started working on a generic check for invalid tags (using TagInfo data and OSM Wiki Data items), and I learned a whole new meaning for "long processing time" (for reference, my 5 country test runs went from ~2 hours to 14+ hours, and in one case, almost 23 hours). I've gotten that back down to ~3 hours, and I'm looking for more optimizations for that, but this isn't in the same category at all.